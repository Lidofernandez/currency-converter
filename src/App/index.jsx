import React, { useState, useEffect } from "react";
import CurrencyField from "../CurrencyField";

const API_CURRENCY_URL = "https://api.exchangeratesapi.io/latest";

const App = () => {
  const [currencies, setCurrencies] = useState([]);
  const [loading, setLoading] = useState(false);

  const [currency, setCurrency] = useState();
  const [value, setValue] = useState(1);
  const [toCurrency, setToCurrency] = useState();
  const [exchangeRate, setExchangeRate] = useState(1);
  const [isValueToConvert, setIsValueToConvert] = useState(true);

  useEffect(() => {
    async function fetchCurrencyRates() {
      try {
        setLoading(true);
        const response = await fetch(API_CURRENCY_URL);

        const data = await response.json();
        setLoading(false);
        const rateList = Object.keys(data.rates);
        setCurrencies([data.base, ...rateList]);
        setCurrency(data.base);
        setToCurrency(rateList[0]);
        setExchangeRate(data.rates[rateList[0]]);
      } catch (error) {
        setLoading(false);
      }
    }

    fetchCurrencyRates();
  }, []);

  useEffect(() => {
    async function fetchCurrencyRate() {
      try {
        setLoading(true);
        const response = await fetch(
          `${API_CURRENCY_URL}?base=${currency}&symbols=${toCurrency}`,
        );

        const data = await response.json();
        setLoading(false);
        setExchangeRate(data.rates[toCurrency]);
      } catch (error) {
        setLoading(false);
      }
    }

    if (currency && toCurrency) {
      if (currency === toCurrency) {
        setExchangeRate(1);
        return;
      }

      fetchCurrencyRate();
    }
  }, [currency, toCurrency]);

  const getCurrency = (event) => {
    setCurrency(event.currentTarget.value);
  };

  const getToCurrency = (event) => {
    setToCurrency(event.currentTarget.value);
  };

  const getValue = (event) => {
    setValue(event.currentTarget.value);
    setIsValueToConvert(true);
  };

  const getToValue = (event) => {
    setValue(event.currentTarget.value);
    setIsValueToConvert(false);
  };

  let valueToConvert;
  let valueConverted;
  if (isValueToConvert) {
    valueToConvert = value;
    valueConverted = value * exchangeRate;
  } else {
    valueConverted = value;
    valueToConvert = value / exchangeRate;
  }

  return (
    <div data-testid="App">
      {loading ? (
        "loading..."
      ) : (
        <>
          <CurrencyField
            currencies={currencies}
            currencyTestid="toConvert"
            selectedCurrency={currency}
            getCurrency={getCurrency}
            handleOnChange={getValue}
            value={valueToConvert}
          />
          is equal to
          <CurrencyField
            currencies={currencies}
            currencyTestid="converted"
            selectedCurrency={toCurrency}
            getCurrency={getToCurrency}
            handleOnChange={getToValue}
            value={valueConverted}
          />
        </>
      )}
    </div>
  );
};

export default App;
