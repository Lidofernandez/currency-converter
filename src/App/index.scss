html {
  font-size: 16px;
  font-family: sans-serif, Helvetica, Arial;
}

.app {
  margin: 2rem;
  display: flex;
  flex-direction: column;
  flex-basis: auto;
  text-align: center;
}

.app__separator {
  margin: 0.5rem;
  width: 10%;
}

@media only screen and (min-width: 520px) {
  .app {
    flex-direction: row;
  }
}
