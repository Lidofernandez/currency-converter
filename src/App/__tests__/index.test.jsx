import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";
import App from "../index";

describe("App", () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation((url) => {
      if (url.includes("base=ChuckNorris&symbols=BruceLee")) {
        return Promise.resolve({
          json: () =>
            Promise.resolve({
              rates: { BruceLee: 30 },
              base: "ChuckNorris",
            }),
        });
      }

      if (url.includes("ChuckNorris&symbols=CaptainAmerica")) {
        return Promise.resolve({
          json: () =>
            Promise.resolve({
              rates: { CaptainAmerica: 500 },
              base: "ChuckNorris",
            }),
        });
      }

      if (url.includes("CaptainAmerica&symbols=BruceLee")) {
        return Promise.resolve({
          json: () =>
            Promise.resolve({
              rates: { BruceLee: 5 },
              base: "CaptainAmerica",
            }),
        });
      }

      return Promise.resolve({
        json: () =>
          Promise.resolve({
            rates: { BruceLee: 10, CaptainAmerica: 500 },
            base: "ChuckNorris",
          }),
      });
    });
  });

  afterEach(() => {
    fetch.mockClear();
  });

  it("should render", async () => {
    const { getByTestId, asFragment } = render(<App />);
    await waitFor(() => getByTestId("App"));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the currencies", async () => {
    const { getByDisplayValue } = render(<App />);
    await waitFor(() => {
      getByDisplayValue("BruceLee");
      getByDisplayValue("ChuckNorris");
    });
  });

  it("should convert from target input", async () => {
    const { getByDisplayValue, getByTestId } = render(<App />);
    await waitFor(() => {
      getByTestId("toConvert");
      getByTestId("converted");
    });

    const chuckNorris = getByDisplayValue("1");
    fireEvent.change(chuckNorris, { target: { value: 9 } });
    await waitFor(() => getByDisplayValue(270));
  });

  it("should convert from ouput input", async () => {
    const { getByDisplayValue, getByTestId } = render(<App />);
    await waitFor(() => {
      getByTestId("toConvert");
      getByTestId("converted");
    });

    const bruceLee = getByDisplayValue("1");
    fireEvent.change(bruceLee, { target: { value: 100 } });
    await waitFor(() => getByDisplayValue(3000));
  });

  it("should convert from target dropdown", async () => {
    const { getByDisplayValue, getByTestId } = render(<App />);
    await waitFor(() => {
      getByTestId("toConvert");
    });

    fireEvent.change(getByTestId("toConvert"), {
      target: { value: "CaptainAmerica" },
    });

    await waitFor(() => getByDisplayValue(5));
    expect(global.fetch).toHaveBeenCalledTimes(3);
  });

  it("should convert from ouput dropdown", async () => {
    const { getByDisplayValue, getByTestId } = render(<App />);
    await waitFor(() => {
      getByTestId("converted");
    });

    fireEvent.change(getByTestId("converted"), {
      target: { value: "CaptainAmerica" },
    });

    await waitFor(() => getByDisplayValue(500));
    expect(global.fetch).toHaveBeenCalledTimes(3);
  });

  it("should not trigger extra API for same options", async () => {
    const { getAllByDisplayValue, getByTestId } = render(<App />);
    await waitFor(() => {
      getByTestId("toConvert");
      getByTestId("converted");
    });

    fireEvent.change(getByTestId("converted"), {
      target: { value: "ChuckNorris" },
    });

    await waitFor(() => getAllByDisplayValue(1));
    expect(global.fetch).toHaveBeenCalledTimes(2);
  });
});
