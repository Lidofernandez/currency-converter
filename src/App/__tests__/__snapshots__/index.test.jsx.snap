// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<DocumentFragment>
  <div
    data-testid="App"
  >
    <div>
      <div>
        <input
          type="number"
          value="1"
        />
        <select
          data-testid="toConvert"
        >
          <option
            value="ChuckNorris"
          >
            ChuckNorris
          </option>
          <option
            value="BruceLee"
          >
            BruceLee
          </option>
          <option
            value="CaptainAmerica"
          >
            CaptainAmerica
          </option>
        </select>
      </div>
      <div>
        =
      </div>
      <div>
        <input
          type="number"
          value="30"
        />
        <select
          data-testid="converted"
        >
          <option
            value="ChuckNorris"
          >
            ChuckNorris
          </option>
          <option
            value="BruceLee"
          >
            BruceLee
          </option>
          <option
            value="CaptainAmerica"
          >
            CaptainAmerica
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
